riscv/src/register/macros.rs

```diff
@@ -600,3 +600,431 @@ macro_rules! clear_pmp {
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $field_ty:ident: [$field_start:literal : $field_end:literal],
+    ) => {
+        paste::paste! {
+            $crate::read_only_csr_field!(
+                $ty,
+                $(#[$field_doc])+
+                $field,
+                $field_ty: [$field_start : $field_end],
+            );
+
+            $crate::write_only_csr_field!(
+                $ty,
+                $(#[$field_doc])+
+                [<set_ $field>],
+                $field_ty: [$field_start : $field_end],
+            );
+        }
+    };
+}
+
+/// Defines field accesor functions for a read-only CSR type.
+#[macro_export]
+macro_rules! read_only_csr_field {
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident: $bit:literal$(,)?) => {
+        const _: () = assert!($bit < usize::BITS);
+
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&self) -> bool {
+                $crate::bits::bf_extract(self.bits, $bit, 1) != 0
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident: $bit_start:literal..=$bit_end:literal$(,)?) => {
+        const _: () = assert!($bit_end < usize::BITS);
+        const _: () = assert!($bit_start < $bit_end);
+
+        paste::paste! {
+            impl $ty {
+                $(#[$field_doc])+
+                #[inline]
+                pub fn $field(&self, index: usize) -> bool {
+                    self.[<try_ $field>](index).unwrap()
+                }
+
+                $(#[$field_doc])+
+                #[inline]
+                pub fn [<try_ $field>](&self, index: usize) -> $crate::result::Result<bool> {
+                    if ($bit_start..=$bit_end).contains(&index) {
+                        Ok($crate::bits::bf_extract(self.bits, index, 1) != 0)
+                    } else {
+                        Err($crate::result::Error::IndexOutOfBounds {
+                            index,
+                            min: $bit_start,
+                            max: $bit_end,
+                        })
+                    }
+                }
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident: [$bit_start:literal : $bit_end:literal]$(,)?) => {
+        const _: () = assert!($bit_end < usize::BITS);
+        const _: () = assert!($bit_start < $bit_end);
+
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&self) -> usize {
+                $crate::bits::bf_extract(self.bits, $bit_start, $bit_end - $bit_start + 1)
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $field_ty:ident: [$field_start:literal : $field_end:literal]$(,)?
+    ) => {
+        const _: () = assert!($field_end < usize::BITS);
+        const _: () = assert!($field_start <= $field_end);
+
+        paste::paste! {
```

Comment:
shouldn't we re-export `paste` to allow external users to use this macros? something like `$crate::paste!`

Reply:
Good catch, I added a re-export commit.